pkg/activator/throttler.go

```diff
@@ -0,0 +1,191 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package activator
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	v1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/serving/pkg/queue"
+	"github.com/knative/serving/pkg/reconciler"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	capacityUpdateFailure = "updating capacity failed"
+	OverloadMessage       = "activator overload"
+)
+
+// ThrottlerParams defines the parameters of the Throttler
+type ThrottlerParams struct {
+	BreakerParams queue.BreakerParams
+	Logger        *zap.SugaredLogger
+	GetEndpoints  func(RevisionID) (int32, error)
+	GetRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+}
+
+// NewThrottler creates a new Throttler
+func NewThrottler(params ThrottlerParams) *Throttler {
+	breakers := make(map[RevisionID]*queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: params.BreakerParams, logger: params.Logger, getEndpoints: params.GetEndpoints, getRevision: params.GetRevision}
+}
+
+// Throttler keeps the mapping of Revisions to Breakers
+// and allows updating max concurrency dynamically of respective Breakers.
+// Max concurrency is essentially the number of semaphore tokens the Breaker has in rotation.
+// The manipulation of the parameter is done via `UpdateCapacity()` method.
+// It enables the use case to start with max concurrency set to 0 (no requests are sent because no endpoints are available)
+// and gradually increase its value depending on the external condition (e.g. new endpoints become available)
+type Throttler struct {
+	breakers      map[RevisionID]*queue.Breaker
+	breakerParams queue.BreakerParams
+	logger        *zap.SugaredLogger
+	getEndpoints  func(RevisionID) (int32, error)
+	getRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+	mux           sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, err := t.updateCapacity(rev, size)
+	return err
+}
+
+// This method updates Breaker's concurrency and requires external synchronization, e.g. use the mux.
+// It create a new breaker and saves it into our bookkeeping if doesn't exist.
+// This is important for not loosing the update signals
+// that came before the requests reached the Activator's Handler.
+func (t *Throttler) updateCapacity(rev RevisionID, size int32) (breaker *queue.Breaker, err error) {
+	revision, exists, err := t.getRevision(rev)
+	if err != nil {
+		return breaker, err
```

Comment:
I feel that an empty pointer is also fine since the error is there and it is usually checked first. Don't have a strong feeling about that, can change it if you see it as more readable.

Reply:
I personally prefer passing nil specifically, that takes out any wondering if we're actually passing a half-initialized object back accidentally.